micronota/bfillings/diamond.py

```diff
@@ -176,85 +179,140 @@ def make_db(in_fp, out_fp=None, params=None):
+            self.run_view(daa_fp, out_fp, params={'--outfmt': outfmt})
+            res = res.append(self.parse_tabular(out_fp))
+            found.extend(res.index)
+            # save to a tmp file the seqs that do not hit current database
+            new_fp = join(self.tmp_dir, '%s.fa' % out_prefix)
+            with open(new_fp, 'w') as f:
+                for seq in read(fp, format='fasta'):
+                    if seq.metadata['id'] not in found:
+                        seq.write(f, format='fasta')
+            # no seq left
+            if stat(new_fp).st_size == 0:
+                break
+            else:
+                fp = new_fp
+        return res
+
+    def run_blast(self, fp, daa_fp, db, aligner='blastp', evalue=0.001, cpus=1,
+                  params=None):
+        '''Search query sequences against the database.
+
+        Parameters
+        ----------
+        fp : str
+            File path for the query sequence.
+        daa_fp : str
+            Output file path.
+        cpus : int
+            Number of CPUs. Default to 1. If it is set to 0, it will use
+            all available CPUs.
+        evalue : float
+            Default to 0.01. Threshold E-value.
+        params : dict
+            Other command line parameters for diamond blastp. key is the option
+            (e.g. "-T") and value is the value for the option (e.g. "50").
+            If the option is a flag, set the value to None.
+
+        Returns
+        -------
+        str
+            The file path of the blast result.
+        '''
+        logger = getLogger(__name__)
+
+        if aligner == 'blastp':
+            app = DiamondBlastp
+        elif aligner == 'blastx':
+            app = DiamondBlastx
+        else:
+            raise ValueError('Unknown aligner: %s.' % aligner)
+
+        blast = app(InputHandler='_input_as_paths', params=params)
+        blast.Parameters['--query'].on(fp)
+        blast.Parameters['--daa'].on(daa_fp)
+        blast.Parameters['--db'].on(db)
+        blast.Parameters['--evalue'].on(evalue)
+        blast.Parameters['--threads'].on(cpus)
+        blast.Parameters['--tmpdir'].on(self.tmp_dir)
+        logger.info('Running: %s' % blast.BaseCommand)
+        blast_res = blast()
+        blast_res.cleanUp()
+        return blast_res
+
+    def run_view(self, daa_fp, out_fp, params=None):
+        '''
+        Parameters
+        ----------
+        daa_fp : str
+            Input file resulting from diamond blast.
+        out_fp : str
+            Output file.
+        '''
+        logger = getLogger(__name__)
+        view = DiamondView(InputHandler='_input_as_paths')
+        view.Parameters['--daa'].on(daa_fp)
+        view.Parameters['--out'].on(out_fp)
+        logger.info('Running: %s' % view.BaseCommand)
+        view_res = view()
+        view_res.cleanUp()
+        return view_res
+
+    @staticmethod
+    def parse_tabular(diamond_res, column='bitscore'):
```

Comment:
What do you think about making bitscore optional?

I can think of scenarios where it may be interesting to return all of the hits rather than the best ones.